File: examples/lh-server/python/08-exception-handling/tasks.py
import random

from littlehorse.exceptions import LHTaskException


async def charge_credit_card(user_id: str, amount: float) -> None:
    if amount > 10000:
        raise LHTaskException(
            "insufficient-funds", f"User {user_id} has insufficient funds"
        )
    if random.randint(0, 1):
        raise RuntimeError("Uh oh, network failure!")
    print(f"Successfully charged credit card of user {user_id}")


async def fetch_amount(user_id: str) -> float:
    return random.uniform(0.0, 100.0)


async def ship_item(item_id: str, user_id: str) -> None:
    print(f"Successfully shipped item {item_id} to user {user_id}")


async def cancel_order_insufficient_funds(user_id: str) -> None:
    print(
        f"Notifying user {user_id} that order was cancelled due to insufficient funds"
    )


async def notify_order_failed(user_id: str) -> None:
    print(f"Notifying user {user_id} that order failed for technical reasons")

File: examples/lh-server/go/08-exception-handling/tasks.go
package main

import (
	"fmt"
	"math/rand"

	"github.com/littlehorse-enterprises/littlehorse/sdk-go/littlehorse"
)

func ChargeCreditCard(userID string, amount float64) (string, error) {
	currentBalance := rand.Float64() * 100
	if amount > currentBalance {
		return "", &littlehorse.LHTaskException{
			Name:    "insufficient-funds",
			Message: fmt.Sprintf("User %s has insufficient funds", userID),
		}
	}
	if rand.Intn(2) == 0 {
		return "", fmt.Errorf("network failure")
	}
	fmt.Printf("Successfully charged credit card of user %s\n", userID)
	return "success", nil
}

func ShipItem(itemID, userID string) {
	fmt.Printf("Successfully shipped item %s to user %s\n", itemID, userID)
}

func CancelOrderInsufficientFunds(userID string) {
	fmt.Printf("Order canceled for %s: insufficient funds\n", userID)
}

func NotifyOrderFailed(userID string) {
	fmt.Printf("Order failed for %s: technical failure\n", userID)
}
